### ISSUE TITLE
[HOLD for payment 2024-04-09] [Held Requests] Approving/Paying expense reports with held requests

### ISSUE THREAD

**Issue by robertjchen** [Expensify/App#31345]
![image](https://github.com/Expensify/App/assets/2438855/ed779fb9-1f41-4607-8ae7-9645b7f92b8a)

![image](https://github.com/Expensify/App/assets/2438855/ff1f92c2-9aa6-4635-863a-911fc3a847a5)

### Paying expense reports with held requests (admin)
When paying out an expense report, we need to total up and present held and non-held amounts.

Under src/components create a new `<ProcessMoneyRequestHoldMenu`. This component will solely contain a `<PopoverMenu` with `iou.confirmPay` (or `iou.confirmApproval`) as the header text, `iou.confirmPayAmount` (or iou.confirmApprovalAmount) as the description body depending on `props.type`, followed by two `<Button>`'s one for the held amount and payable amount. The two props the menu takes are: 
- `props.fullAmount`
- `props.heldAmount`
- `props.type` - either 'pay' or 'approve'

This component is only displayed if the payer can unhold the requests, which we'll determine with the function below.

In `src/components/ReportActionItem/ReportPreview.js`, update `<SettlementButton` to first call a new local method `confirmSettle()`, we will move the existing `IOU.payMoneyRequest` into this new local method.

The new local method would do the following:
1. Total up all reimbursable requests on the report and determine if there are any held amounts within that total
2. If not, directly call `IOU.payMoneyRequest` 
3. Otherwise, check if the user has permissions to unhold the requests
4. If not, directly call `IOU.payMoneyRequest`
5. If they can, make the new `<PayMoneyRequestHoldMenu` visible, passing the full and held amounts to it so that it may be displayed on the buttons.

In `src/libs/actions/IOU.js` update `payMoneyRequest()` to take a new additional flag full which will determine whether or not we pay the full amount, or just the held amounts. This flag will also be passed to the backend API as part of the call.

In the case where we reveal the  `<PayMoneyRequestHoldMenu`, the user has the option to pay the full amount or just the non-held amount. Depending on what they select, we'll either call `IOU.payMoneyRequest` with the `full` parameter set to true or false.

### Approving expense reports with held requests (approvers)
Similar to the above, we'll make the same changes but to different places:

In `src/components/ReportActionItem/ReportPreview.js`, update `<Button`  under `shouldShowApproveButton` to first call a new local method `confirmApproval()`, we will move the existing `IOU.approveMoneyRequest` into this new local method.

In the new local method, check for held amounts, permissions and call `IOU.approveMoneyRequest()` as appropriate (same logic as above).

Pass 'approve' as the type parameter to the `<ProcessMoneyRequestHoldMenu` component.

The rest should be the same, particularly with adding a new full parameter to `IOU.payMoneyRequest` and passing that onwards into the backend.

**Comment by BartoszGrajdek** [comment]
Hello, I'm Bartosz Grajdek from Software Mansion, the expert agency, and I would like to work on this task.

**Comment by BartoszGrajdek** [comment]
There's a mistake in the task description I think
> followed by two Button's one for the held amount and payable amount

Shouldn't it say `followed by two <Button>'s one for the not held amount and one for payable amount`?

**Comment by BartoszGrajdek** [comment]
Also in:
> Approving expense reports with held requests (approvers)
In src/components/ReportActionItem/ReportPreview.js, update <Button under shouldShowApproveButton

I think that the file you meant was actually `src/components/MoneyReportHeader`

**Comment by robertjchen** [comment]
You are correct, apologies for the confusion! 🙇

**Comment by robertjchen** [comment]
In progress!

**Comment by robertjchen** [comment]
In progress!

**Comment by BartoszGrajdek** [comment]
After a discussion with Robert, it turns out that some things weren't mentioned in the issue description, but they are necessary for it to work properly. Working on them right now and discussing everything with Robert accordingly 🙏🏻

**Comment by robertjchen** [comment]
Working with Bartosz to iron out a few final items in the PR: https://github.com/Expensify/App/pull/33124 as well as make corresponding backend changes.

**Comment by robertjchen** [comment]
We're working on getting hold functionality out first, and will raise discussion around offline flows/final items around this shortly.

**Comment by robertjchen** [comment]
Please see main issue for a full-picture update: https://github.com/Expensify/Expensify/issues/274076

**Comment by robertjchen** [comment]
In progress under WIP PR: https://github.com/Expensify/App/pull/33124

**Comment by robertjchen** [comment]
Please see https://github.com/Expensify/Expensify/issues/274076#issuecomment-1899437428 for the latest consolidated update!

**Comment by johnmlee101** [comment]
Just FYI added @allroundexperts to do the C+ review checklist of that PR ^

**Comment by trjExpensify** [comment]
Yo yo! @robertjchen @JmillsExpensify @BartoszGrajdek... just a heads up, I'm going to go through and prefix these project issues with `[Held requests]` because I can filter on that for the #wave-collect project board, but equally, a lot of people using `in:title: [Hold]` to filter out issues on hold.

**Comment by robertjchen** [comment]
Make sense, thanks! I've added it to all the other [related GH issues](https://expensify.slack.com/archives/C02NZ2HGJAZ/p1709337801198499) from this list as well 👍

**Comment by robertjchen** [comment]
Latest update: https://expensify.slack.com/archives/C036QM0SLJK/p1710490163098179?thread_ts=1710203436.990009&cid=C036QM0SLJK

- Marco was able to reproduce the issue with the partial pay/approve flow and has a [draft PR](https://github.com/Expensify/Auth/pull/10221/files) up.

**Comment by robertjchen** [comment]
Latest update [with more context on Slack](https://expensify.slack.com/archives/C036QM0SLJK/p1710815560310969?thread_ts=1710800386.718709&cid=C036QM0SLJK).

**Comment by robertjchen** [comment]
- Backend bugs were resolved and the changes were merged and deployed 🙌 
- [FE PR](https://github.com/Expensify/App/pull/33124) now unblocked, final [Onyx-related](https://github.com/Expensify/Web-Expensify/pull/41362) changes in progress. Hopefully, we can get these merged by EOW? 🤞

**Comment by bondydaa** [comment]
I had this blocker https://github.com/Expensify/App/issues/39335 created which I think is related to this PR https://github.com/Expensify/App/pull/33124.

The "bug" is that when you click approve on the request preview in the "main chat" it doesn't trigger the hold confirmation model like it does when you're directly viewing the manual request. I don't actually see that mentioned in the OP though so it might just not have been considered initially. 

Is this something we'd want to block the roll out of the feature for? I'm leaning "no" but just want to double check that with ya'll.

**Comment by bondydaa** [comment]
okay coolio I'll remove the label then, want me to just assign the issue to @BartoszGrajdek and/or you? or let it go through the normal external bug flow?

**Comment by robertjchen** [comment]
I'd say external would be good!

**Comment by robertjchen** [comment]
**Update**

The FE changes are on staging and we've had a few bug reports come through already:
- [No confirmation prompt when approving held request in the main chat](https://github.com/Expensify/App/issues/39335) - proposals by contributor under review
- [Pay with Expensify button does not include amount after approving held request](https://github.com/Expensify/App/issues/39342) - proposals by contributors already lined up
- [Confirmation to pay is still shown after approving all the requests](https://github.com/Expensify/App/issues/39336) - Appears to be frontend issue, should be able to be handled by contributors

Keep those reports coming! :raised_hands:
